File: Algo_Coursera-Specialization/AlgoToolbox/week4_divide_and_conquer/3_improving_quicksort/sorting.py
# Uses python3
import sys
import random


def partition3(a, l, r):
    # write your code here
    x = a[l]
    j = l+1
    k = l

    for i in range(l + 1, r + 1):
        if a[i] <= x:
            k += 1
            a[i], a[k] = a[k], a[i]
            if a[k] < x:
                a[j], a[k] = a[k], a[j]
                j += 1

    a[l], a[j-1] = a[j-1], a[l]

    return [j, k]


def partition2(a, l, r):
    x = a[l]
    j = l
    for i in range(l + 1, r + 1):
        if a[i] <= x:
            j += 1
            a[i], a[j] = a[j], a[i]
    a[l], a[j] = a[j], a[l]
    return j


def randomized_quick_sort(a, l, r):
    if l >= r:
        return
    k = random.randint(l, r)
    a[l], a[k] = a[k], a[l]
    # use partition3
    [m1, m2] = partition3(a, l, r)
    randomized_quick_sort(a, l, m1 - 1)
    randomized_quick_sort(a, m2 + 1, r)


if __name__ == '__main__':
    n = int(input())
    a = list(map(int, input().split()))
    randomized_quick_sort(a, 0, n - 1)
    for x in a:
        print(x, end=' ')

File: Algo_Coursera-Specialization/AlgoToolbox/week4_divide_and_conquer/3_improving_quicksort/sorting.cpp
#include <cstdlib>
#include <iostream>
#include <vector>

using std::swap;
using std::vector;

int partition2(vector<int> &a, int l, int r) {
    int x = a[l];
    int j = l;
    for (int i = l + 1; i <= r; i++) {
        if (a[i] <= x) {
            j++;
            swap(a[i], a[j]);
        }
    }
    swap(a[l], a[j]);
    return j;
}

vector<int> partition3(vector<int> &a, int l, int r) {
    int x = a[l];
    vector<int> m(2);
    int j = l;
    int k = r;
    int i = l;
    while (i <= k) {
        if (a[i] < x) {
            swap(a[j], a[i]);
            j++;
            i++;
        } else if (a[i] == x) {
            i++;
        } else {
            swap(a[i], a[k]);
            k -= 1;
        }
        m[0] = j;
        m[1] = k;
    }
    return m;
}
void randomized_quick_sort(vector<int> &a, int l, int r) {
    if (l >= r) {
        return;
    }

    int k = l + rand() % (r - l + 1);
    swap(a[l], a[k]);
    int m1 = partition3(a, l, r)[0];
    int m2 = partition3(a, l, r)[1];

    randomized_quick_sort(a, l, m1 - 1);
    randomized_quick_sort(a, m2 + 1, r);
}

int main() {
    int n;
    std::cin >> n;
    vector<int> a(n);
    for (size_t i = 0; i < a.size(); ++i) {
        std::cin >> a[i];
    }
    randomized_quick_sort(a, 0, a.size() - 1);
    for (size_t i = 0; i < a.size(); ++i) {
        std::cout << a[i] << ' ';
    }
}
